scvi/model/_condscvi.py

```diff
@@ -128,34 +128,79 @@ def get_vamp_prior(
         )
         key = labels_state_registry.original_key
         mapping = labels_state_registry.categorical_mapping
-        for ct in range(self.summary_stats.n_labels):
-            # pick p cells
-            local_indices = np.random.choice(
-                np.where(adata.obs[key] == mapping[ct])[0], p
+
+        scdl = self._make_data_loader(adata=adata, batch_size=p)
+
+        mean = []
+        var = []
+        for tensors in scdl:
+            x = tensors[REGISTRY_KEYS.X_KEY]
+            y = tensors[REGISTRY_KEYS.LABELS_KEY]
+            out = self.module.inference(x, y)
+            mean_, var_ = out["qz_m"], out["qz_v"]
+            mean += [mean_.cpu()]
+            var += [var_.cpu()]
+
+        mean_cat, var_cat = np.array(torch.cat(mean)), np.array(torch.cat(var))
+        adata.obsm["X_CondSCVI"] = mean_cat
+
+        for ct in range(self.summary_stats["n_labels"]):
+            local_indices = np.where(adata.obs[key] == mapping[ct])[0]
+            sub_adata = adata[local_indices, :].copy()
+            if "overclustering_vamp" not in sub_adata.obs.columns:
+                try:
+                    from scanpy.pp import neighbors
+                    from scanpy.tl import leiden
+                except ImportError:
+                    raise ImportError(
+                        """
+                            Please install leidenalg and scanpy package via `pip install scanpy leidenalg`
+                            or provide results of overclustering in adata.obs["overclustering_camp"]
+                        """
+                    )
+                neighbors(sub_adata, use_rep="X_CondSCVI")
+                leiden(
+                    sub_adata, resolution=resolution, key_added="overclustering_vamp"
+                )
+
+            var_cluster = np.zeros(
+                [
+                    len(np.unique(sub_adata.obs.overclustering_vamp)),
+                    self.module.n_latent,
+                ]
+            )
+            mean_cluster = np.zeros(
+                [
+                    len(np.unique(sub_adata.obs.overclustering_vamp)),
```

Comment:
len(np.unique(sub_adata.obs.overclustering_vamp)) <- give this a proper name. It's used twice and long

Reply:
refactor